File: dynamic_programming/coin_change.kt
fun main() {
    val s = Solution()
    val coins = intArrayOf(1,3,5,11,17)
    val amount = 21
    println(s.coinChange(coins, amount))
    println(s.coinChange_dp(coins, amount))

}

class Solution {

    fun coinChange_dp(coins: IntArray, amount: Int): Int {
        // Dynamic programming solution
        // O(n*m) where n = amount, m = coins.size
        if (coins.isEmpty() || amount < 0) {
            return -1
        }

        if (amount == 0) {
            return 0
        }

        // f(i) is the min number of coins to reach amount i
        val dp = IntArray(amount+1) {0}

        // base cases
        for (c in coins) {
            if (c < dp.size) {
                dp[c] = 1
            }
        }

        // f(i) = min(f(i-coins[0]), ..., f(i-coins[n])) + 1
        for (i in 2..amount) {
            var curMin = Int.MAX_VALUE
            for (c in coins) {

                // current coin is greater than the amount we're aiming for
                if (c > i) continue

                // if adding c to the last amount gets us at least to amount
                // then it's a candidate for the new minimum
                if ((dp[i-c]+1)*c >= i) {
                    curMin = minOf(curMin, dp[i-c])
                }
            }

            // if we found a new minimum, memoize it plus one
            if (curMin < Int.MAX_VALUE) {
                dp[i] = curMin + 1
            }
        }

        // return the last memoized item, or -1 if no coins can get us to amount
        return if (dp.last() > 0) dp.last() else -1
    }


    fun coinChange(coins: IntArray, amount: Int): Int {
        // O(n*m*k) - where n = amount, m = coins.size, k = number of unique sums of coins
        if (coins.isEmpty() || amount < 0) {
            return -1
        }

        if (amount == 0) {
            return 0
        }

        var prev = mutableSetOf<Int>(0)
        for (i in 1..amount) {
            var cur = mutableSetOf<Int>()
            for (j in prev) {
                for (c in coins) {
                    val curSum = j + c
                    if (curSum == amount) {
                        // we found the mininum number of coins!
                        return i
                    }
                    if (curSum < amount) {
                        cur.add(curSum)
                    }
                }
            }
            prev = cur
        }

        // cannot make amount out of coins
        return -1
    }
}

File: dynamic_programming/coin_change.py
#! /usr/bin/env python3

class Solution:
    def coinChange(self, coins: list[int], amount: int) -> int:
        if not coins or amount < 0:
            return -1

        if amount == 0:
            return 0

        coin_set = set(coins)
        # we can bail early if there is a coin that matches amount
        if amount in coin_set:
            return 1

        # memo[i] = minimum number of coins to make i
        # base case: 1 if there is a coin of value i
        memo = [1 if x in coin_set else 0 for x in range(amount+1)]

        for i in range(2, amount+1):
            cur_min = amount+1
            for c in coins:
                if c > i:
                    continue
                if (memo[i-c]+1)*c >= i:
                    cur_min = min(cur_min, memo[i-c])

            if cur_min < amount + 1:
                memo[i] = cur_min + 1

        return -1 if memo[-1] == 0 else memo[-1]

    def coinChange2(self, coins: list[int], amount: int) -> int:
        if not coins or amount < 0:
            return -1

        memo = [amount+1] * (amount+1)
        memo[0] = 0
        for c in coins:
            for i in range(c, amount+1):
                memo[i] = min(memo[i], memo[i-c]+1)

        if memo[-1] == amount+1:
            return -1
        return memo[-1]


if __name__ == "__main__":
    s = Solution()
    coins = [2, 5]
    amount = 3
    print(s.coinChange(coins, amount))
